pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -522,7 +522,7 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.Pip
 	resources.ApplyTaskResults(nextRprts, resolvedResultRefs)
 
 	for _, rprt := range nextRprts {
-		if rprt == nil {
+		if rprt == nil || rprt.Skip(pipelineState, d) {
```

Comment:
This looks fine for now but would like to see this being evaluated as an attribute to `WhenExpressions` similar to what we have for the parameter. That way, all the checks are in one place irrespective of param mismatch or task result mismatch, a task and all its dependencies are guaranteed to be skipped.

Reply:
the point where we excluded the tasks with when expressions containing static inputs or parameter variables (which are replaced earlier as read from the pipeline) is here: https://github.com/tektoncd/pipeline/blob/079a6c8bce6eb5663689fad7f50d0fce7591e6bc/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L260-L272


at that point, when expressions containing results don't have the variables replaced so they are not excluded -- that function is used here: 
https://github.com/tektoncd/pipeline/blob/906414229d999cd9c7e6a5401f937c0bb3d97e7f/pkg/reconciler/pipelinerun/pipelinerun.go#L499-L511


the task results are applied later right before this additional check to exclude the when expressions containing results: https://github.com/tektoncd/pipeline/blob/906414229d999cd9c7e6a5401f937c0bb3d97e7f/pkg/reconciler/pipelinerun/pipelinerun.go#L522-L527

the options I see are:

1. removing the earlier check, so that we have only one check after task results have been applied -- but this means will be processing more tasks than necessary which could have been excluded earlier
2. as we have it now, do two checks with the first one reducing the list next tasks and reducing unnecessary work -- I'm leaning towards this but can see the benefit of doing the check once

please let me know what you think

Reply:
@jerop I am with you - option 2 is fine.

But I was thinking more towards, adding call to `ApplyTaskResults` in `when_types.go` and update `AllowsExecution` to deal with constants, params, and task results. `AllowsExecution` in the end would be responsible for checking if constants match, resolving params/task results and check if they match. Similar to what we do with `applyReplacements` and `GetVarSubsutitionExpressions` as attributes to `WhenExpression`.

These changes does not have to be part of this PR. It's more about refactoring which can be done in a separate PR.

Reply:
thanks for the clarification @pritidesai -- will do that refactor soon